File: template.go
package snd

import (
	"fmt"
	"strings"
)

type TemplateConfig struct {
	Key         string      `json:"key"`
	Name        string      `json:"name"`
	Description string      `json:"description"`
	Type        string      `json:"type"`
	Default     interface{} `json:"default"`
}

// Template represents one S&D template.
type Template struct {
	Name          string                 `json:"name"`
	Slug          string                 `json:"slug"`
	Author        string                 `json:"author"`
	Description   string                 `json:"description"`
	PrintTemplate string                 `json:"printTemplate"`
	ListTemplate  string                 `json:"listTemplate"`
	SkeletonData  map[string]interface{} `json:"skeletonData"`
	Images        map[string]string      `json:"images"`
	Config        []TemplateConfig       `json:"config"`
	DataSources   []string               `json:"dataSources"`
	Version       string                 `json:"version"`
}

func (t Template) ID() string {
	return fmt.Sprintf("tmpl:%s+%s", t.Author, t.Slug)
}

func IsTemplateID(id string) bool {
	return strings.HasPrefix(id, "tmpl:")
}

File: frontend/src/js/types/template.ts
import { pickBy } from 'lodash-es';

import BasicInfo from 'js/types/basic-info';
import { ConfigValue } from 'js/types/config';

type Template = BasicInfo & {
	printTemplate: string;
	listTemplate: string;
	skeletonData: Record<string, any>;
	images: Record<string, string>;
	config: ConfigValue[];
	dataSources: string[];
	count?: number;
};

/**
 * Sanitizes the config object by setting default values for missing fields and removing old fields that are not present in the config anymore.
 * @param t Template object.
 * @param configs Config object.
 */
const sanitizeConfig = (t: Template, configs: any) => {
	// Create base config
	if (configs === undefined) {
		configs = {};
	}

	// Set default values for initialized fields
	(t.config ?? []).forEach((conf) => {
		if (configs[conf.key] === undefined) {
			configs[conf.key] = conf.default;
		}
	});

	// Remove old fields that are not present in the config anymore.
	return pickBy(configs, (val, key) => key === 'seed' || t.config.some((conf) => conf.key === key));
};

/**
 * Creates an empty template object.
 */
function createEmptyTemplate(): Template {
	return {
		name: 'Your Template Name',
		description: '',
		copyrightNotice: '',
		author: 'username',
		slug: 'your-template-name',
		version: '',
		printTemplate: '',
		listTemplate: '',
		skeletonData: {},
		images: {},
		config: [],
		dataSources: [],
	};
}

export default Template;
export { createEmptyTemplate, sanitizeConfig };
